workspace/resource_notebook.go

```diff
@@ -139,6 +143,86 @@ func (a NotebooksAPI) Mkdirs(path string) error {
 	}, nil)
 }
 
+type syncAnswer struct {
+	MU   sync.Mutex
+	data []ObjectStatus
+}
+
+func (a *syncAnswer) append(objs []ObjectStatus) {
+	a.MU.Lock()
+	a.data = append(a.data, objs...)
+	a.MU.Unlock()
+}
+
+type directoryInfo struct {
+	Path     string
+	Attempts int
+}
+
+const directyListingMaxAttempts = 3
+
+func (a NotebooksAPI) recursiveAddPathsParallel(directory directoryInfo, dirChannel chan directoryInfo,
+	answer *syncAnswer, wg *sync.WaitGroup) {
+	defer wg.Done()
+	notebookInfoList, err := a.list(directory.Path)
+	if err != nil {
+		log.Printf("[WARN] error listing '%s': %v", directory.Path, err)
+		if directory.Attempts < directyListingMaxAttempts {
+			wg.Add(1)
+			dirChannel <- directoryInfo{Path: directory.Path, Attempts: directory.Attempts + 1}
+		}
+	}
+	answer.append(notebookInfoList)
+	for _, v := range notebookInfoList {
+		if v.ObjectType == Directory {
+			wg.Add(1)
+			log.Printf("[DEBUG] %s: putting directory '%s' into channel. Channel size: %d",
+				time.Now().Local().Format(time.RFC3339Nano), v.Path, len(dirChannel))
+			dirChannel <- directoryInfo{Path: v.Path}
+			time.Sleep(15 * time.Millisecond)
+		}
+	}
+}
+
+func (a NotebooksAPI) ListParallel(path string, recursive bool) ([]ObjectStatus, error) {
+	var answer syncAnswer
+	wg := &sync.WaitGroup{}
+
+	numWorkers := 5
+	if os.Getenv("EXPORTER_WS_LIST_PARALLELISM") != "" {
+		t, _ := strconv.ParseInt(os.Getenv("EXPORTER_WS_LIST_PARALLELISM"), 10, 32)
+		numWorkers = int(t)
+	}
+	channelSize := 100000
+	if os.Getenv("EXPORTER_CHANNEL_SIZE") != "" {
+		t, _ := strconv.ParseInt(os.Getenv("EXPORTER_CHANNEL_SIZE"), 10, 32)
+		channelSize = int(t)
+	}
```

Comment:
I think abstracting this logic outside might be good since we are duplicating getting the os.Getenv()
````
func getEnvAsInt(envName string, defaultValue int) int {
	if val, exists := os.LookupEnv(envName); exists {
		parsedVal, err := strconv.Atoi(val)
		if err == nil {
			return parsedVal
		}
	}
	return defaultValue
}
````